File: 09-embedding/golang/main.go
package main

import (
	"fmt"
  "log"
	wasm "github.com/wasmerio/go-ext-wasm/wasmer"
)

func main() {
	bytes, err := wasm.ReadBytes("my_lib.wasm")
  if err != nil {
    log.Fatalf("error reading wasm: %s", err)
  }
	instance, err := wasm.NewInstance(bytes)
  if err != nil {
    log.Fatalf("error instantiating wasm: %s", err)
  }
	defer instance.Close()

	sum := instance.Exports["sum_of_squares"]
	result, err := sum(100)
  if err != nil {
    log.Fatalf("error calling wasm library: %s", err)
  }

	fmt.Println(result)
}

File: 09-embedding/rust/src/main.rs
use wasmtime::*;

fn main() {
  let engine = HostRef::new(Engine::default());
  let store = HostRef::new(Store::new(&engine));

  let raw_bytes = std::fs::read("my_lib.wasm").unwrap();
  let module = HostRef::new(Module::new(&store, &raw_bytes).unwrap());
  let instance = Instance::new(&store, &module, &[]).unwrap();

  let sum_of_squares = instance.find_export_by_name("sum_of_squares").unwrap().func().unwrap();

  let result = sum_of_squares.borrow().call(&[Val::I32(100)]).unwrap();
  println!("{:?}", result);
}
